### PR TITLE
JS enhancements for "Add another" further conditions

### PR DESCRIPTION
**Pull request by benswannack**
## Context
We added the ability to add and remove further conditions previously. This required refreshing the page each time though. This PR adds progressive JS enhancements for that feature to make filling out this form more smooth.

## Changes proposed in this pull request
Add JS enhancements to further conditions form
Add JS system spec for the feature

## Guidance to review
To test: Go though the flow in the review app. Try this with JS enabled and disabled and ensure the behaviour is the same

## Link to Trello card
https://trello.com/c/u0fZBMtg/3613-implement-js-progressive-enhancements-to-infinite-offer-conditions

## Things to check

- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] This code does not rely on the addition/removal of Azure config environment variables in the same Pull Request
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/docs/environment-variables.md#azure-hosting-devops-pipeline)

### CODE REVIEW

File: app/frontend/packs/further_conditions/add_another.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import assignIndexBasedValues from './helper'
+import removeFurtherCondition from './remove'
+
+const addFurtherConditionElement = () => {
+  const items = document.getElementsByClassName('app-add-another__item')
```
Comment (steventux) [inline app/frontend/packs/further_conditions/add_another.js:5]: Is `.app-add-another__item` descriptive enough for the entire document? It seems a bit generic. Perhaps `.app-add-condition*` as a convention?

Also should the base string(s) for these repeatedly defined class names be in an exported constant somewhere?

```js
const CSS_CLASS_ROOT = 'app-add-condition'
const ITEM_CSS_CLASS = CSS_CLASS_ROOT + '__item'
...
```

**Review by steventux — APPROVED** [review]
Looks good and works well, a couple of very minor nitpicks.

File: app/frontend/packs/further_conditions/add_another.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  const latestId = items.length
+  const newNode = buildNewFurtherConditionField(latestId)
+  document.querySelector('.app-add-another__add-button').insertAdjacentElement('beforebegin', newNode)
+  newNode.querySelector('textarea').focus()
```
Comment (cwrw) [inline app/frontend/packs/further_conditions/add_another.js:9]: can we limit the add another to 18 conditions (now its infinite)?

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -135,7 +135,7 @@ jobs:
         run: bundle exec rake db:setup
 
       - name: ${{ matrix.tests }} tests with feature flags ${{ matrix.feature-flags }}
-        run: bundle exec --verbose rspec --exclude-pattern ${EXCLUDE_PATTERN} --pattern ${INCLUDE_PATTERN} --format documentation
+        run: apk add chromium chromium-chromedriver && bundle exec --verbose rspec --exclude-pattern ${EXCLUDE_PATTERN} --pattern ${INCLUDE_PATTERN} --format documentation
```
Comment (cwrw) [inline .github/workflows/build.yml:138]: Can we move adding those to a different step? Also Im wondering around this slowing down our builds, Github actions has caching capabilities, is it possible to extend them for this step

File: app/frontend/packs/further_conditions/add_another.js (hunk 2 above)
Comment (benswannack) [reply app/frontend/packs/further_conditions/add_another.js:9]: Ah good spot

File: spec/support/capybara.rb [hunk 4]
```diff
@@ -5,8 +5,26 @@
   Capybara.server_port = 9887 + ENV['TEST_ENV_NUMBER'].to_i
 end
 
+Capybara.register_driver :chrome_headless do |app|
```
Comment (benswannack) [inline spec/support/capybara.rb:8]: Copying over from the other review as still relevant:

The default selenium chrome headless driver doesn't work in Docker. See https://stackoverflow.com/questions/50610316/capybara-headless-chrome-in-docker-returns-devtoolsactiveport-file-doesnt-exist

File: spec/support/webmock.rb [hunk 5]
```diff
@@ -1 +1,3 @@
 require 'webmock/rspec'
+
+WebMock.disable_net_connect!(allow_localhost: true)
```
Comment (benswannack) [inline spec/support/webmock.rb:3]: If we don't set this, then the chrome headless driver has its requests blocked. Allowing just localhost seems safe as we won't bombard any actual services with test requests

**Review by cwrw — COMMENTED** [review]
Well done, especially with breaking ground on system JS specs and setting up! Some minor comments but all works as expected 👍

File: .github/workflows/build.yml (hunk 3 above)
Comment (benswannack) [reply .github/workflows/build.yml:138]: Good point, as discussed I'm looking into github actions caching

File: app/frontend/packs/further_conditions/index.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
+import addFurtherConditionElement from './add_another'
```
[the remark this answers is not in the document]
Comment (benswannack) [reply app/frontend/packs/further_conditions/index.js]: I've moved some stuff around to split up the add/remove logic a bit better

File: app/frontend/packs/further_conditions/updateAddButtonVisibility.js [hunk 7]
```diff
@@ -0,0 +1,15 @@
+import { ADD_BUTTON_CSS_CLASS, ITEM_CSS_CLASS } from './constants'
+
+const MAX_FURTHER_CONDITIONS = 18
```
Comment (benswannack) [inline app/frontend/packs/further_conditions/updateAddButtonVisibility.js:3]: Not entirely happy with this.

Maybe we could render this hidden in the form html somewhere and read from it in the JS?

File: app/frontend/packs/further_conditions/add_another.js (hunk 2 above)
Comment (benswannack) [reply app/frontend/packs/further_conditions/add_another.js]: Fixed and added JS unit test to cover

File: .github/workflows/build.yml (hunk 3 above)
Comment (benswannack) [reply .github/workflows/build.yml]: I've moved it to a different step.

Not sure about the caching though - I was looking to borrow from other projects, but they don't cache chromedriver (e.g. https://github.com/DFE-Digital/register-trainee-teachers/blob/master/.github/workflows/build.yml#L59) 

It takes 8s to do the step, maybe we can revisit caching in another PR?

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -134,6 +134,9 @@ jobs:
       - name: Setup Database
         run: bundle exec rake db:setup
 
+      - name: Install chromedriver
+        run: apk add chromium chromium-chromedriver
```
Comment (benswannack) [inline .github/workflows/build.yml:138]: Need to update the README to mention the new dependency on chromedriver to run the specs

File: app/frontend/packs/further_conditions/updateAddButtonVisibility.js (hunk 7 above)
Comment (cwrw) [reply app/frontend/packs/further_conditions/updateAddButtonVisibility.js:3]: Yea maybe in a data hash in hidden html, if we were using `js.erb` partials would have been easier but we aren't here so I would go with the first option

**Review by cwrw — COMMENTED** [review]
One last thing I remembered is API conditions vs Manage conditions. I think we can do this in a followup PR as keen to ship this, but we need to catch the validation error when submitting an offer with over 20 conditions. This can be replicated by setting 20 further conditions, and then editing the form conditions and selecting the extra standard conditions and attempting to change the offer.

**Comment by benswannack** [comment]
> One last thing I remembered is API conditions vs Manage conditions. I think we can do this in a followup PR as keen to ship this, but we need to catch the validation error when submitting an offer with over 20 conditions. This can be replicated by setting 20 further conditions, and then editing the form conditions and selecting the extra standard conditions and attempting to change the offer.

I've raised a bug card for that, we can pick up after

File: .github/workflows/build.yml (hunk 3 above)
Comment (cwrw) [reply .github/workflows/build.yml]: sure, as there is scope for caching multiple steps

**Review by cwrw — APPROVED** [review]
Thanks @benswannack 👏  can we ping the tech channel around the new requirement just as an FYI

**Pull request merged**